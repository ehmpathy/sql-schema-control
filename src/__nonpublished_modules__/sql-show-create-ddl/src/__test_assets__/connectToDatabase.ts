import mysql from 'mysql2/promise';
import pg from 'pg';

import { ConnectionConfig, DatabaseLanguage } from '../../../../types';
import { DatabaseConnection } from '../types';

// create the connection to database
const connectionAdapters = {
  [DatabaseLanguage.MYSQL]: async ({
    connectionConfig,
  }: {
    connectionConfig: ConnectionConfig;
  }): Promise<DatabaseConnection> => {
    const connection = await mysql.createConnection({
      host: connectionConfig.host,
      port: connectionConfig.port,
      database: connectionConfig.schema,
      user: connectionConfig.username,
      password: connectionConfig.password,
      multipleStatements: true,
    });
    return {
      query: async ({ sql, values }: { sql: string; values?: any[] }) => {
        const result = await connection.query(sql, values);
        return { rows: result[0] }; // standardize the response
      },
      end: async () => connection.end(),
    };
  },
  [DatabaseLanguage.POSTGRES]: async ({
    connectionConfig,
  }: {
    connectionConfig: ConnectionConfig;
  }): Promise<DatabaseConnection> => {
    const client = new pg.Client({
      host: connectionConfig.host,
      port: connectionConfig.port,
      user: connectionConfig.username,
      password: connectionConfig.password,
      database: connectionConfig.schema,
    });
    await client.connect();
    return {
      query: async ({ sql, values }: { sql: string; values?: (string | number)[] }) => {
        const result = await client.query(sql, values);
        return { rows: result.rows };
      },
      end: () => client.end(),
    };
  },
};

export const connectToDatabase = async ({
  language,
  config,
}: {
  language: DatabaseLanguage;
  config: ConnectionConfig;
}): Promise<DatabaseConnection> => {
  // 1. get the connection adapter for the method
  const getDbConnection = connectionAdapters[language];

  // 2. attempt to connect
  const connection = await getDbConnection({ connectionConfig: config });

  // 3. run a test query to check connection
  await connection.query({ sql: 'SELECT 1' });

  // 4. return the connection
  return connection;
};
