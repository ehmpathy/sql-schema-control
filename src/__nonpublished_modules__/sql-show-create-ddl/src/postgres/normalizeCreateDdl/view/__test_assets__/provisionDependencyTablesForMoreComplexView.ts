import { DatabaseConnection } from '../../../../types';

const dependencyTablesDdl = [
  `
CREATE TABLE home (
  id bigserial NOT NULL,
  uuid uuid NOT NULL,
  created_at timestamp with time zone NOT NULL DEFAULT now(),
  name varchar(255) NOT NULL,
  built timestamp with time zone NOT NULL,
  bedrooms int NOT NULL,
  bathrooms int NOT NULL,
  host_ids_hash bytea NOT NULL,
  CONSTRAINT home_pk PRIMARY KEY (id),
  CONSTRAINT home_ux1 UNIQUE (name, host_ids_hash)
);
      `.trim(),
  `
CREATE TABLE home_version (
  id bigserial NOT NULL,
  home_id bigint NOT NULL,
  effective_at timestamp with time zone NOT NULL DEFAULT now(),
  created_at timestamp with time zone NOT NULL DEFAULT now(),
  photo_ids_hash bytea NOT NULL,
  CONSTRAINT home_version_pk PRIMARY KEY (id),
  CONSTRAINT home_version_ux1 UNIQUE (home_id, effective_at, created_at),
  CONSTRAINT home_version_fk0 FOREIGN KEY (home_id) REFERENCES home (id)
);
CREATE INDEX home_version_fk0_ix ON home_version USING btree (home_id);
      `.trim(),
  `
CREATE TABLE home_version_to_photo (
  id bigserial NOT NULL,
  created_at timestamp with time zone NOT NULL DEFAULT now(),
  home_version_id bigint NOT NULL,
  photo_id bigint NOT NULL,
  array_order_index smallint NOT NULL,
  CONSTRAINT home_version_to_photo_pk PRIMARY KEY (id),
  CONSTRAINT home_version_to_photo_ux1 UNIQUE (home_version_id, array_order_index),
  CONSTRAINT home_version_to_photo_fk0 FOREIGN KEY (home_version_id) REFERENCES home_version (id)
  -- CONSTRAINT home_version_to_photo_fk1 FOREIGN KEY (photo_id) REFERENCES photo (id)
);
CREATE INDEX home_version_to_photo_fk0_ix ON home_version_to_photo USING btree (home_version_id);
CREATE INDEX home_version_to_photo_fk1_ix ON home_version_to_photo USING btree (photo_id);
      `.trim(),
  `
CREATE TABLE home_to_host (
  id bigserial NOT NULL,
  created_at timestamp with time zone NOT NULL DEFAULT now(),
  home_id bigint NOT NULL,
  host_id bigint NOT NULL,
  array_order_index smallint NOT NULL,
  CONSTRAINT home_to_host_pk PRIMARY KEY (id),
  CONSTRAINT home_to_host_ux1 UNIQUE (home_id, array_order_index),
  CONSTRAINT home_to_host_fk0 FOREIGN KEY (home_id) REFERENCES home (id)
  -- CONSTRAINT home_to_host_fk1 FOREIGN KEY (host_id) REFERENCES host (id)
);
CREATE INDEX home_to_host_fk0_ix ON home_to_host USING btree (home_id);
CREATE INDEX home_to_host_fk1_ix ON home_to_host USING btree (host_id);
      `.trim(),
  `
CREATE TABLE home_cvp (
  id bigserial NOT NULL,
  updated_at timestamp with time zone NOT NULL DEFAULT now(),
  home_id bigint NOT NULL,
  home_version_id bigint NOT NULL,
  CONSTRAINT home_cvp_pk PRIMARY KEY (id),
  CONSTRAINT home_cvp_ux1 UNIQUE (home_id),
  CONSTRAINT home_cvp_fk0 FOREIGN KEY (home_id) REFERENCES home (id),
  CONSTRAINT home_cvp_fk1 FOREIGN KEY (home_version_id) REFERENCES home_version (id)
);
CREATE INDEX home_cvp_fk0_ix ON home_cvp USING btree (home_id);
CREATE INDEX home_cvp_fk1_ix ON home_cvp USING btree (home_version_id);
      `.trim(),
];
const dependencyTablesNames = [
  'home_cvp',
  'home_to_host',
  'home_version_to_photo',
  'home_version',
  'home',
];
export const provisionDependencyTablesForMoreComplexView = async ({
  dbConnection,
}: {
  dbConnection: DatabaseConnection;
}) => {
  await Promise.all(
    dependencyTablesNames.map((tableName) =>
      dbConnection.query({ sql: `DROP TABLE IF EXISTS ${tableName} CASCADE` }),
    ),
  );
  await Promise.all(
    dependencyTablesDdl.map((ddl) => dbConnection.query({ sql: ddl })),
  );
};
