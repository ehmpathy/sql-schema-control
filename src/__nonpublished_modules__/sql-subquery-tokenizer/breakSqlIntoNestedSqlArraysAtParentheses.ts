type ValueOrArray<T> = T | ValueOrArray<T>[];
export type NestedStringArray = ValueOrArray<string>[];

const castNewlinesAndQuotesIntoJSONParsableNewlinesAndQuotes = (str: string) =>
  str.replace(/\n/g, '\\n').replace(/"/g, '\\"');
export const breakSqlIntoNestedSqlArraysAtParentheses = ({
  sql,
}: {
  sql: string;
}): NestedStringArray => {
  const sqlWithNewlinesReplaced =
    castNewlinesAndQuotesIntoJSONParsableNewlinesAndQuotes(sql);
  const sqlWithLeftParensReplaced = sqlWithNewlinesReplaced.replace(
    /\(/g,
    '",["(',
  );
  const sqlWithLeftAndRightParensReplaced = sqlWithLeftParensReplaced.replace(
    /\)/g,
    ')"],"',
  );
  const jsonArrayOfParenNestedSql = `["${sqlWithLeftAndRightParensReplaced}"]`;
  const arrayOfParenNestedSql: NestedStringArray = JSON.parse(
    jsonArrayOfParenNestedSql,
  ); // use json to break it up for us; inspiration: https://stackoverflow.com/a/40478000/3068233
  return arrayOfParenNestedSql;
};
