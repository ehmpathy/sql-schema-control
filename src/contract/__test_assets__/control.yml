language: mysql
dialect: 5.7
connection: ./connection.config.js
strict: false # strict = true -> display uncontrolled resources
definitions:
  - ./definitions/tables.yml
  - ./definitions/functions.yml
  - ./definitions/procedures.yml
  - ./definitions/init.yml
