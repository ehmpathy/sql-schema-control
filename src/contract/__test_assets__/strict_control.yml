language: mysql
dialect: 5.7
connection: ./connection.config.js
strict: true # strict = true -> display uncontrolled resources
definitions:
  - ./definitions/tables.yml
  - ./definitions/views.yml
  - ./definitions/functions.yml
  - ./definitions/procedures.yml
  - ./definitions/init.yml
