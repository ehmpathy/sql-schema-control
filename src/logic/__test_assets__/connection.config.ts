// note: this connection config is defined in /provision/integration_test_db/docker-compose
export const promiseConfig = async () => ({
  mysql: {
    host: 'localhost',
    port: 12821,
    schema: 'superimportantdb',
    username: 'root',
    password: 'a-secure-password',
  },
  postgresql: {
    host: 'localhost',
    port: 7821,
    schema: 'superimportantdb',
    username: 'postgres',
    password: 'a-secure-password',
  },
});
