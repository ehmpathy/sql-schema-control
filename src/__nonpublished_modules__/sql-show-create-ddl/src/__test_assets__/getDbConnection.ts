import { DatabaseLanguage } from '../../../../domain';
import { connectToDatabase } from './connectToDatabase';

// the connection config for our integration test dbs, provisioned by docker
export const config = {
  [DatabaseLanguage.MYSQL]: {
    host: 'localhost',
    port: 12821,
    database: 'superimportantdb',
    username: 'root',
    password: 'a-secure-password',
  },
  [DatabaseLanguage.POSTGRES]: {
    host: 'localhost',
    port: 7821,
    database: 'superimportantdb',
    schema: 'public',
    username: 'postgres',
    password: 'a-secure-password',
  },
};

export const getDbConnection = async ({
  language,
}: {
  language: DatabaseLanguage;
}) => connectToDatabase({ language, config: config[language] });
