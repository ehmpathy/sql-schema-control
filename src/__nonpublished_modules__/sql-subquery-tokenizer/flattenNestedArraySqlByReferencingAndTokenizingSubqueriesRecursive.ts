import { NestedStringArray } from './breakSqlIntoNestedSqlArraysAtParentheses';
import { extractAndTokenizeSubqueryReferencesInArrayOfSqlStrings } from './extractAndTokenizeSubqueryReferencesInArrayOfSqlStrings';
import { SqlSubqueryReference } from './model/SqlSubqueryReference';

export const flattenNestedArraySqlByReferencingAndTokenizingSubqueriesRecursive =
  ({
    sqlOrNestedSqlArray,
  }: {
    sqlOrNestedSqlArray: NestedStringArray | string;
  }): { references: SqlSubqueryReference[]; flattenedSql: string } => {
    // 0. if this is a string, then no references - its already flat
    if (!Array.isArray(sqlOrNestedSqlArray)) {
      return {
        references: [],
        flattenedSql: sqlOrNestedSqlArray,
      };
    }

    // 1. since its not a string, its a nested sql array; update name of const for readability
    const nestedSqlArray = sqlOrNestedSqlArray;

    // 2. initialize the references object, so we can track references over nested arrays
    const references: SqlSubqueryReference[] = [];

    // 3. for each element in this array, flatten it. track the nested references
    const flattenedSqlParts = nestedSqlArray.map((thisSqlOrNestedSqlArray) => {
      const { references: nestedReferences, flattenedSql: nestedFlattenedSql } =
        flattenNestedArraySqlByReferencingAndTokenizingSubqueriesRecursive({
          sqlOrNestedSqlArray: thisSqlOrNestedSqlArray,
        });
      references.push(...nestedReferences);
      return nestedFlattenedSql;
    });

    // 4. extract references from this array of flattened strings
    const { references: referencesFromFlattenedSqlParts, referencedSqlParts } =
      extractAndTokenizeSubqueryReferencesInArrayOfSqlStrings({
        sqlParts: flattenedSqlParts,
      });
    references.push(...referencesFromFlattenedSqlParts); // append the newly found references

    // 5. join all of the flattened, referenced strings
    const flattenedSql = referencedSqlParts.join('');

    // 6. return the references and sql
    return {
      references,
      flattenedSql,
    };
  };
