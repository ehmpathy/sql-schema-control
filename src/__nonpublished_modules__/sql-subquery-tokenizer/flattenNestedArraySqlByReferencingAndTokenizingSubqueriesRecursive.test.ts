import { exampleSqlQueryWithSubquery } from './__test_assets__/exampleSqlQueryWithSubquery';
import { breakSqlIntoNestedSqlArraysAtParentheses } from './breakSqlIntoNestedSqlArraysAtParentheses';
import { flattenNestedArraySqlByReferencingAndTokenizingSubqueriesRecursive } from './flattenNestedArraySqlByReferencingAndTokenizingSubqueriesRecursive';

import uuid = require('uuid');
jest.mock('uuid');
const uuidMock = uuid as any as jest.Mock;
uuidMock.mockReturnValue('__UUID__');

describe('flattenNestedArraySqlByReferencingAndTokenizingSubqueriesRecursive', () => {
  it('should accurately flatten and tokenize this example', async () => {
    const sql = exampleSqlQueryWithSubquery;
    const nestedSqlArray = breakSqlIntoNestedSqlArraysAtParentheses({ sql });
    const { references, flattenedSql } =
      flattenNestedArraySqlByReferencingAndTokenizingSubqueriesRecursive({
        sqlOrNestedSqlArray: nestedSqlArray,
      });
    expect(references.length).toEqual(1);
    expect(references).toMatchSnapshot();
    expect(flattenedSql).toMatchSnapshot();
  });
});
