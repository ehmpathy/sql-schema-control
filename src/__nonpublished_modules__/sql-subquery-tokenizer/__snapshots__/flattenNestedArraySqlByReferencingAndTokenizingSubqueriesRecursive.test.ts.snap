// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`flattenNestedArraySqlByReferencingAndTokenizingSubqueriesRecursive should accurately flatten and tokenize this example 1`] = `
Array [
  SqlSubqueryReference {
    "id": "__UUID__",
    "sql": "(
    SELECT GROUP_CONCAT(ice_cream_to_ingredient.ingredient_id ORDER BY ice_cream_to_ingredient.ingredient_id)
    FROM ice_cream_to_ingredient WHERE ice_cream_to_ingredient.ice_cream_id = s.id
  )",
  },
]
`;

exports[`flattenNestedArraySqlByReferencingAndTokenizingSubqueriesRecursive should accurately flatten and tokenize this example 2`] = `
"SELECT
  s.id,
  s.uuid,
  s.name,
  __SSQ:__UUID____ as ingredient_ids,
  s.created_at
FROM ice_cream s
;"
`;
